### PR TITLE
new uninlined_format_args lint to inline explicit arguments

### PR DESCRIPTION
**Pull request by nyurik**
Implement https://github.com/rust-lang/rust-clippy/issues/8368 - a new lint to inline format arguments such as `print!("{}", var)` into `print!("{var}")`.

### Supported cases

code | suggestion | comment
---|---|---
`print!("{}", var)` | `print!("{var}")` |  simple variables
`print!("{0}", var)` | `print!("{var}")` |  positional variables
`print!("{v}", v=var)` | `print!("{var}")` |  named variables
`print!("{0} {0}", var)` | `print!("{var} {var}")` |  aliased variables
`print!("{0:1$}", var, width)` | `print!("{var:width$}")` |  width support
`print!("{0:.1$}", var, prec)` | `print!("{var:.prec$}")` |  precision support
`print!("{:.*}", prec, var)` | `print!("{var:.prec$}")` |  asterisk support

### Known Problems

* There may be a false positive if the format string is wrapped in a macro call:
```rust
# let var = 42;
macro_rules! no_param_str { () => { "{}" }; }
macro_rules! pass_through { ($expr:expr) => { $expr }; }
println!(no_param_str!(), var);
println!(pass_through!("{}"), var);
```

* Format string uses an indexed argument that cannot be inlined.
Supporting this case requires re-indexing of the format string.
Until implemented, `print!("{0}={1}", var, 1+2)` should be changed to `print!("{var}={0}", 1+2)` by hand.

changelog: [`uninlined_format_args`]: A new lint to inline format arguments, i.e. `print!("{}", var)` into `print!("{var}")`

### CODE REVIEW

**Review by Alexendoo — COMMENTED** [review]
👍

I had a look through most of the `TODO`s. Also worth mentioning you can use [let chains](https://github.com/rust-lang/rust/pull/94927) rather than `if_chain`, it plays better with rust_analyzer (though also doesn't get automatically formatted yet)

**Comment by nyurik** [comment]
Thanks @Alexendoo for the review!  I made a number of changes based on that, but still a few todos.  My biggest concern is how to detect if an argument is also used as precision and/or width.  I know some work is being done in the rustc, and there seem to be at least 3 devs involved, so I wonder who/how can coordinate these efforts to be consistent and reuse each other's work.

**Comment by est31** [comment]
If it's not in your TODOs yet, the lint should be integrated into the MSRV system of clippy and be gated on 1.58.0, as [that's when format capturing was added](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1580-2022-01-13).

**Comment by nyurik** [comment]
@est31 thx, fixed.  I think I am done with this lint -- it now solves the majority of cases, and it avoids expanding width/precision arguments by checking if the argument span contains a `$` symbol.  Not ideal, but much better than nothing.

**Comment by xFrednet** [comment]
@Alexendoo Do you want to take over the full review? I believe you know the format macro better than me. If you don't have the time, feel free to give it back. I should have more time for Clippy again :upside_down_face:

**Comment by Alexendoo** [comment]
Sure! Sounds good to me. I'll take a full look at it once it's unblocked

**Comment by xFrednet** [comment]
Awesome, thank you!

r? @Alexendoo

**Comment by nyurik** [comment]
Thx @Alexendoo !  Is there something i can do to help the #8518 ?  Are there any fundamental blocks on that, or is it juts a matter of patiently waiting? :)

**Comment by Alexendoo** [comment]
Yeah just a matter of waiting for the large amount of reviewer bandwidth needed

File: clippy_lints/src/inline_format_args.rs
Comment (flip1995) [inline clippy_lints/src/inline_format_args.rs:24]: The lint name is wrong. It suggests that this **lints on** inline format args rather than that it **suggests** using them.

Maybe `needless_explicit_format_args`? :thinking:

**Review by flip1995 — COMMENTED** [review]
Thanks for this PR! I learned something about the format macro and what is possible with it.

The impl LGTM overall, but some housekeeping, like documentation, has to be done still.

File: clippy_utils/src/macros.rs
Comment (nyurik) [reply clippy_utils/src/macros.rs:578]: @flip1995 argument can be empty, but it still exists:   `format!("{}", foo)` is an empty argument span, but its **position** is there.  This is different from width and precision, which might be omitted (optional).

File: tests/ui/inline_format_args.rs
Comment (nyurik) [reply tests/ui/inline_format_args.rs:45]: @flip1995 I could, but with the recent changes to the underlying format string parsing code, it might be possible for me to handle more cases - which SHOULD be inlined, but I didn't have the needed info to handle them.

File: clippy_lints/src/inline_format_args.rs
Comment (nyurik) [reply clippy_lints/src/inline_format_args.rs:24]: ah, naming... :)  Maybe `explicit_format_args` is a bit simpler?  or `uninlined_format_args` / `noninlined_format_args`?

**Comment by nyurik** [comment]
@flip1995 thanks for the review! I addressed some of your points, and made a few clarifications. I am still in a bit of a limbo state - waiting for the upstream #8518 conflicts to be resolved, plus now there is #9349 - so I am not certain at which point I should start resolving conflicts and getting it back to the "ready to merge" state again.

File: clippy_utils/src/macros.rs
Comment (flip1995) [reply clippy_utils/src/macros.rs]: Ah so it is the span between the `{}`, excluding the `{}`? If so, that wasn't immediately clear to me from the documentation.

File: clippy_lints/src/inline_format_args.rs
Comment (flip1995) [reply clippy_lints/src/inline_format_args.rs:28]: I'm not the best at lint naming. Let's decide this right before merging. We can also ask other on Zulip by then.

**Comment by flip1995** [comment]
No worries. Let's get the other PRs merged first. You can work on this as you see fit. But also feel free to just let this PR sit as-is until we resolved the other issues.

**Comment by nyurik** [comment]
I decided to spend a bit of time rewriting inlining -- now it's part of the same `format_args.rs` as the other similar lints, and uses the new parser.

New functionality:  now supports aliases, width, and precision argument inlining, including the starred syntax (eeek!).

While at it, I discovered that my unit tests detected a regression #9468.  In this PR, I simply revert the offending patch, but I hope it will get fixed before merging.

One big TODO: handle the positional references, and renumber them if needed.

**Comment by nyurik** [comment]
This PR is now rebased on main branch, ready for review. The biggest TODO is what should we name this thing?  The current name suggests that this lints on inline format args rather than that it suggests using them (per @flip1995)

Suggestions so far:
* needless_explicit_format_args
* explicit_format_args
* uninlined_format_args
* noninlined_format_args
* ???

**Comment by est31** [comment]
My suggestion: `inlineable_format_args`.

From the ones you suggested @nyurik `explicit_format_args` is my favourite.

Also there is the question about which things to abbreviate. Should it be `args` or `arguments`? Should it be `fmt` or `format`?

**Comment by nyurik** [comment]
Looking at all the lints, there are no `FMT`, only `FORMAT`. There is a mix of `ARG` (6) and `ARGUMENT` (2), so it seems the `ARG` variant is more common.

**Comment by est31** [comment]
Hmm yeah, the macro in std is also called `format_args`, so I guess writing out `format` and shortening `args` is the best way.

**Comment by nyurik** [comment]
Seems it is very common to use `needless` in the lint names, so per Lukas (zulip) suggestion, renaming it to `needless_format_args` for now.

File: clippy_utils/src/macros.rs
Comment (Alexendoo) [inline clippy_utils/src/macros.rs:730]: oops, thanks 😄

File: clippy_lints/src/format_args.rs
Comment (Alexendoo) [inline clippy_lints/src/format_args.rs:173]: This is pretty useful, I think it'd be worth adding to `FormatParam` (maybe with a different name so there aren't both `kind` and `type` fields), that way you could have like `for param in format_args.params() { check_inline(..) }`

**Review by Alexendoo — COMMENTED** [review]
Thanks for waiting :D

File: tests/ui/needless_format_args.rs
Comment (Alexendoo) [reply tests/ui/needless_format_args.rs:21]: or `let _: &[u8; 3] = b"[tab] [tab]";` for the current form

File: clippy_lints/src/format_args.rs
Comment (nyurik) [reply clippy_lints/src/format_args.rs:190]: do you want me to apply some change to macros.rs?  I'm not certain which it is from the diff.  Or do you want to submit it separatelly?

File: clippy_lints/src/format_args.rs
Comment (nyurik) [reply clippy_lints/src/format_args.rs:173]: thx, yeah, makes things cleaner

File: clippy_lints/src/format_args.rs
Comment (nyurik) [reply clippy_lints/src/format_args.rs:132]: dope, i thought `Vec::new()` always allocates memory, even if empty.  Should have read the docs.

File: clippy_lints/src/format_args.rs
Comment (Alexendoo) [reply clippy_lints/src/format_args.rs:188]: You can do it as part of this PR

Comment (nyurik) [reply clippy_lints/src/format_args.rs:188]: yes, but which part?  Apply your diff?

Comment (nyurik) [reply clippy_lints/src/format_args.rs:188]: @Alexendoo I tried applying your change, and it did indeed makes it a bit cleaner, thanks!

File: clippy_lints/src/format_args.rs
Comment (Alexendoo) [inline clippy_lints/src/format_args.rs:156]: nit: this comment should be a lil further down, it doesn't apply to the MSRV stuff

File: clippy_utils/src/macros.rs
Comment (Alexendoo) [inline clippy_utils/src/macros.rs:566]: A doc comment on these would be nice, something along the lines of `/// Appears as an argument, e.g. {param}` and similar for width/precision

File: tests/ui/needless_format_args.fixed
Comment (nyurik) [reply tests/ui/needless_format_args.fixed:123]: @Alexendoo I tried to add a test like you suggested, but for some strange reason it is not working (the .fixed file shows both as expanded).  It seems the `FormatArgs` is instantiated with `msrv=None`. Is there some other magical thing I have to do somewhere to ensure that clippy gets the right msrv?

**Review by nyurik — COMMENTED** [review]
thx @Alexendoo , addressed all your feedback, but ran into an issue with msrv -- see my other comment

File: clippy_lints/src/format_args.rs
Comment (Alexendoo) [inline clippy_lints/src/format_args.rs:118]:
```suggestion
    msrv: Option<RustcVersion>,
```

+ other changes as needed, moving the `meets_msrv(..)` to the `check_expr`

File: clippy_lints/src/format_args.rs
Comment (Alexendoo) [inline clippy_lints/src/format_args.rs:175]:
```suggestion
    }
    
    extract_msrv_attr!(LateContext);
```

**Review by Alexendoo — COMMENTED** [review]
The MSRV changes:

Also needed is to add the lint name to https://github.com/rust-lang/rust-clippy/blob/e120fb10c60a27670c72e7ec99e93d16c045f493/clippy_lints/src/utils/conf.rs#L216

File: clippy_lints/src/format_args.rs
Comment (nyurik) [reply clippy_lints/src/format_args.rs:173]: Aha!  This is the piece i was missing. Writing code with macros... oh well. Thanks!!!

**Comment by nyurik** [comment]
Thanks @Alexendoo !  The extract_msrv_attr! maagic did the trick.

**Comment by Alexendoo** [comment]
I looked through some `lintcheck` output and it looks good, but multiline formats run into https://github.com/rust-lang/rust/issues/92741 pretty hard:

image

The suggestion is still correct as far as the lint is concerned and `cargo fix` will apply it correctly, the output is just a bit rough

**Comment by nyurik** [comment]
@Alexendoo how can this be fixed? Any suggested changes?

**Comment by nyurik** [comment]
One more thing: when i switched to "style", it caused a lot of uitest errors because many test files use the uninlined formats. Should I fix them, or add a clippy allow to them all?

**Comment by Alexendoo** [comment]
This would all be easier if clippy were a workspace but yeah there's also `clippy_dev` to do

**Comment by nyurik** [comment]
@Alexendoo thanks.  I just discovered a new bug with the code when using the [indoc!](https://crates.io/crates/indoc) crate:

```rust
    println!(indoc!{"foo={}"}, local_i32);
    printdoc!("foo={}", local_i32);
```

The first line is actually inlined:

```
    println!(indoc!{"foo={local_i32}"});
```

indoc uses AST token rewriting, and yet somehow the current format args code manages to inline it. How can I detect such cases, and exclude them?

**Comment by Alexendoo** [comment]
That is a pain, since `indoc!` is a proc macro that is producing a string literal with its span set to its input it's not marked as from expansion, and since it's compatible tokenization wise `clippy_utils::is_from_proc_macro` wouldn't catch it either

**Comment by nyurik** [comment]
@Alexendoo but we must detect this case, otherwise the lint produces uncompilable code :(

See also https://github.com/dtolnay/indoc/issues/50

**Comment by Alexendoo** [comment]
For that issue - as the message explains there's nothing the `indoc` crate can do to make `format!(indoc!())` work, a string coming from a macro can't capture variables

On our end one option would be to tokenize the space between the format string and the first value arg, it should find only whitespace/comments/a comma

**Comment by nyurik** [comment]
@Alexendoo thx, is there an example of how i can do this?  I'm still a bit new to clippy internals

**Comment by nyurik** [comment]
P.S. what should i do about the thousands of `tests/ui/*.rs` files which are not part of any project?  I tried running `cargo dev lint tests/ui/blocks_in_if_conditions.rs -- --fix` hoping that i could just run it for each file in that dir, but it refuses the `--fix` if I use lint with a file instead of a dir.  It would be quiet a task to modify all test files by hand

**Comment by Alexendoo** [comment]
Something along the lines of

```rust
let str_span = format_args.format_string.span;

let between = str
    .with_lo(str_span.hi())
    .with_hi(hygiene::walk_chain(first_value.span, str_span.ctxt()));

if let Some(between_snippet) = snippet_opt(cx, between)
    && tokenize(between_snippet).all(|token| matches!(token.kind, [the comment/comma/whitespace kinds] ))
```

(untested)

**Comment by nyurik** [comment]
@Alexendoo I'm afraid I will need a bit more help with that code.  I tried to implement it, but the span does not capture the needed  code. Plus I am having a very hard time attaching a debugger to the uitest - it keeps stopping at panics in the debug mode.

This is where i'm at:
https://github.com/rust-lang/rust-clippy/pull/9233/files#diff-d77429793a4ce2aaa4372bc9cae50f80c6195c0e46ff9c49358077bff0e60a72R185

Basically i'm not sure what the first value is in your example above. Also, while I think we can test for commas/comments, it might be easier to test for `CloseParen | CloseBrace | CloseBracket`, or even better -- try to detect if there is any token between the outermost span and the format string -- e.g. try to see if format str is wrapper by anything  - indoc in this case: `println!(  indoc!(  "format str" ), ...)`

**Comment by Alexendoo** [comment]
Personally I'm inclined to leave it as a known issue for now. It's not limited to just this lint, any other lint working with string literals may have a problem with this style of proc macro, `print_literal` is affected at least

I plan to replace `expand_past_previous_comma` with something more robust, I can probably catch this issue at the same time with that

**Comment by nyurik** [comment]
@Alexendoo ok, thx, I squashed and cleaned up the PR - should be in a good state now.  I also added a few more unit tests that would also fail until your changes.

File: Cargo.toml
Comment (Alexendoo) [inline Cargo.toml:49]: We can leave this out, don't need to have a test for it currently as the behaviour isn't fixed yet, when it is we'll use an `aux-build` proc macro or just the passthrough `macro_rules` to avoid an extra dependency

File: clippy_lints/src/format_args.rs
Comment (Alexendoo) [inline clippy_lints/src/format_args.rs:166]: You can add a check for `args.format_string.span.from_expansion()`, that'll catch the case where the string comes from a macro (excluding the weird proc macro case)

File: clippy_lints/src/format_args.rs
Comment (nyurik) [reply clippy_lints/src/format_args.rs]: good catch, thx

File: clippy_lints/src/format_args.rs
Comment (nyurik) [reply clippy_lints/src/format_args.rs:166]: @Alexendoo thanks, it does fix the `println!(no_param_str!(), local_i32);` case, where the macro expands into a format string, but it doesn't handle these cases with the pass_through macro (`from_expansion` returns false):

```rust
    macro_rules! pass_through { ($expr:expr) => { $expr }; }
    ...
    println!(pass_through!("foo={local_i32}"), local_i32 = local_i32);
    println!(pass_through!("foo={}"), local_i32);
```

Comment (Alexendoo) [reply clippy_lints/src/format_args.rs:166]: Yeah in that example there's no distinguishing the span of `pass_through!("str")` and `"str"`

Comment (nyurik) [reply clippy_lints/src/format_args.rs:166]: @alexendoo ok, thanks! How would you like to proceed? We could either merge it as a pedantic lint due to false positives, and fix it later, or postpone it and possibly make it as style once fixed?

File: clippy_lints/src/format_args.rs
Comment (Alexendoo) [inline clippy_lints/src/format_args.rs:100]:
```suggestion
    /// format!("{}", var);
    /// format!("{v:?}", v = var);
    /// format!("{0} {0}", var);
    /// format!("{0:1$}", var, width);
    /// format!("{:.*}", prec, var);
    /// ```
    /// Use instead:
    /// ```rust
    /// # let var = 42;
    /// # let width = 1;
    /// # let prec = 2;
    /// format!("{var}");
    /// format!("{var:?}");
    /// format!("{var} {var}");
    /// format!("{var:width$}");
    /// format!("{var:.prec$}");
```

**Review by Alexendoo — COMMENTED** [review]
Looks good! Just some documentation nits and then it's good for merging

Thank you for the endurance 😄

File: clippy_lints/src/format_args.rs
Comment (Alexendoo) [reply clippy_lints/src/format_args.rs:166]: My plan is to merge it, I've kept it waiting long enough 😄

**Comment by Alexendoo** [comment]
Good ol' `update_lints` :D

Could give it one final squash too

**Comment by nyurik** [comment]
Bummer, I can't,  I'm at a playground))))

**Comment by nyurik** [comment]
I'll do it in a bit. Thanks for all the help!!!

**Comment by nyurik** [comment]
@Alexendoo ok, I think it's finally ready 😀

**Review by Alexendoo — APPROVED** [review]

**Comment by Alexendoo** [comment]
Awesome, thanks @nyurik!

@bors r+

**Pull request merged**
